### PR TITLE
More Advanced Matching Algorithm

### PR DESCRIPTION
**Pull request by AniMachaidze**
Add more advanced matching algorithm and test cases.

Algorithm is based on points system which describes similarity between reviewee-reviewer pair. Possible pairs are sorted in decreasing order based on their _matchpoints_. Going through the list, available pairs with biggest _matchpoints_ are matched. People with 0 _matchpoints_ are not matched unless on of them has been waiting for a match for more than 3 days.

### CODE REVIEW

File: resume-buddy/src/main/java/com/google/sps/Match.java [hunk 1]
```diff
@@ -6,12 +6,39 @@
...
+import java.util.concurrent.TimeUnit;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
 
 /** Class that matches reviewers to reviewees */
-public class Match {
+@WebServlet("/match")
+public class Match extends HttpServlet {
```
Comment (oydang) [inline resume-buddy/src/main/java/com/google/sps/Match.java]: Nit: rename this class to MatchServlet and move it to the /servlets folder.

File: resume-buddy/src/main/java/com/google/sps/Match.java [hunk 2]
```diff
@@ -26,31 +53,116 @@ private Match() {}
...
+    for (Entity reviewee : reviewees) {
+      for (Entity reviewer : reviewers) {
+        // TODO: Try different pointing systems to see which works the best
+        int matchPoint = 0;
 
-      Entity matchEntity = new Entity("Match");
+        String revieweeCareer = (String) reviewee.getProperty("career");
+        String reviewerCareer = (String) reviewer.getProperty("career");
+        if (revieweeCareer.equals(reviewerCareer)) {
+          matchPoint += 3;
```
Comment (oydang) [inline resume-buddy/src/main/java/com/google/sps/Match.java]: The points system is a great idea!

You can make these points constants at the top of this class. Eg.
```
private static final int CAREER_POINTS = 3;
private static final int SCHOOL_POINTS = 2;
...
```

File: resume-buddy/src/main/java/com/google/sps/Match.java [hunk 3]
```diff
@@ -26,31 +53,116 @@ private Match() {}
...
+        Date currentDate = new Date();
+        long timeSinceSubmitReviewee = currentDate.getTime() - revieweeSubmitDate.getTime();
+        long timeSinceSubmitReviewer = currentDate.getTime() - reviewerSubmitDate.getTime();
+        timeSinceSubmitReviewee =
+            TimeUnit.DAYS.convert(timeSinceSubmitReviewee, TimeUnit.MILLISECONDS);
+        timeSinceSubmitReviewer =
+            TimeUnit.DAYS.convert(timeSinceSubmitReviewer, TimeUnit.MILLISECONDS);
+
+        // If it has not been 3 days for either of them, then don't match
+        if (timeSinceSubmitReviewee < 3 && timeSinceSubmitReviewer < 3) {
```
Comment (oydang) [inline resume-buddy/src/main/java/com/google/sps/Match.java]: The 3 days should be a constant as well.

`private static final int DAYS_TO_WAIT_UNTIL_TRY_MATCH = 3` (you can come up with a better name than mine, haha).

File: resume-buddy/src/main/java/com/google/sps/Match.java [hunk 4]
```diff
@@ -26,31 +53,116 @@ private Match() {}
...
 
+      // TODO: Send emails to matched people
+
       // Delete reviewers and reviewees from Datastore once matched
       datastore.delete(reviewer.getKey());
       datastore.delete(reviewee.getKey());
     }
   }
+
+  /** Comparator that compares based on the point value and sorts list from bigest to smallest */
```
Comment (oydang) [inline resume-buddy/src/main/java/com/google/sps/Match.java]: Nit: biggest

**Review by oydang — CHANGES_REQUESTED** [review]
Awesome job! Just some stylistic nits.

**Comment by oydang** [comment]
Please resolve merge conflicts first, then I'll be able to review the most up-to-date code!

File: resume-buddy/src/main/java/com/google/sps/Match.java [hunk 5]
```diff
@@ -26,31 +36,117 @@ private Match() {}
...
 
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    int numberOfMatched = Math.min(reviewees.size(), reviewers.size());
 
-    for (int i = 0; i < numberOfMatched; i++) {
-      Entity reviewer = reviewers.get(i);
-      Entity reviewee = reviewees.get(i);
+    // For every reviewee-reviewer pair, count their match point
+    for (Entity reviewee : reviewees) {
+      for (Entity reviewer : reviewers) {
```
Comment (astrelni) [inline resume-buddy/src/main/java/com/google/sps/Match.java:55]: Should probably add a check at the start of the inner loop for whether reviewee and reviewer are the same person.

File: resume-buddy/src/main/java/com/google/sps/Match.java [hunk 6]
```diff
@@ -6,12 +6,22 @@
 import com.google.appengine.api.datastore.FetchOptions;
 import com.google.appengine.api.datastore.PreparedQuery;
 import com.google.appengine.api.datastore.Query;
+import com.google.sps.data.Pair;
+import com.google.sps.data.ReviewStatus;
+import java.util.*;
```
Comment (oydang) [inline resume-buddy/src/main/java/com/google/sps/Match.java]: Google style guide: Avoid wildcard imports. https://google.github.io/styleguide/javaguide.html#s3.3.1-wildcard-imports

File: resume-buddy/src/main/java/com/google/sps/data/ReviewStatus.java [hunk 7]
```diff
@@ -0,0 +1,6 @@
+package com.google.sps.data;
+
+public enum ReviewStatus {
```
Comment (oydang) [inline resume-buddy/src/main/java/com/google/sps/data/ReviewStatus.java:3]: A quick comment explaining how this enum is used would be helpful.

**Review by astrelni — DISMISSED** [review]

File: resume-buddy/src/main/java/com/google/sps/servlets/RevieweeDataServlet.java [hunk 8]
```diff
@@ -67,6 +68,7 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     revieweeEntity.setProperty("career", career);
     revieweeEntity.setProperty("preferred-degree", degreePref);
     revieweeEntity.setProperty("preferred-experience", numYearsPref);
+    revieweeEntity.setProperty("submitdate", new Date());
```
Comment (oydang) [inline resume-buddy/src/main/java/com/google/sps/servlets/RevieweeDataServlet.java]: This should be `submit-date` instead of `submitdate`.

**Review by oydang — CHANGES_REQUESTED** [review]

**Review by oydang — APPROVED** [review]

**Pull request merged**